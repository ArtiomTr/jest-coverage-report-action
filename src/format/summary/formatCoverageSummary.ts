import table from 'markdown-table';

import { CoverageSummary } from '../../typings/Coverage';
import { formatPercentage } from '../../utils/formatPercentage';
import { getStatusOfPercents } from '../../utils/getStatusOfPercents';
import { i18n } from '../../utils/i18n';
import { withExplanation } from '../../utils/withExplanation';

export const formatCoverageSummary = (
    headSummary: Array<CoverageSummary>,
    baseSummary: Array<CoverageSummary> | undefined,
    threshold: number | undefined
): string =>
    table(
        [
            [
                withExplanation(i18n('status'), i18n('statusExplanation')),
                i18n('category'),
                i18n('percentage'),
                i18n('ratio'),
            ],
            ...headSummary.map((currSummary, index) => [
                getStatusOfPercents(currSummary.percentage, threshold),
                currSummary.title,
                formatPercentage(
                    currSummary.percentage,
                    baseSummary?.[index].percentage
                ),
                `${currSummary.covered}/${currSummary.total}`,
            ]),
        ],
        { align: ['c', 'l', 'l', 'c'] }
    );
