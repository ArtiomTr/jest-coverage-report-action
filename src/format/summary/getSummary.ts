import { CoverageSummary } from '../../typings/Coverage';
import { CoverageMap, FileCoverage } from '../../typings/JsonReport';
import { getPercents } from '../getPercents';

export const getSummary = (
    map: CoverageMap,
    totalCounter: (value: FileCoverage) => number,
    coveredCounter: (value: FileCoverage) => number,
    title: string
): CoverageSummary => {
    const total = Object.values(map).reduce(
        (acc, currValue) => acc + totalCounter(currValue),
        0
    );

    const covered = Object.values(map).reduce(
        (acc, currValue) => acc + coveredCounter(currValue),
        0
    );

    return {
        title,
        total,
        covered,
        percentage: getPercents(covered, total),
    };
};
