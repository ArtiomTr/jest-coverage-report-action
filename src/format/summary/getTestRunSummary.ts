import { JsonReport } from '../../typings/JsonReport';
import { i18n } from '../../utils/i18n';

export const getTestRunSummary = (jsonReport: JsonReport) =>
    jsonReport.success
        ? i18n('testsSuccessSummary', {
              numPassedTests: jsonReport.numPassedTests,
              numPassedTestSuites: jsonReport.numPassedTestSuites,
              ending: jsonReport.numPassedTestSuites > 1 ? 's' : '',
          })
        : i18n('testsFailSummary', {
              numFailedTests: jsonReport.numFailedTests,
              numTotalTests: jsonReport.numTotalTests,
              numFailedTestSuites: jsonReport.numFailedTestSuites,
              numTotalTestSuites: jsonReport.numTotalTestSuites,
          });
