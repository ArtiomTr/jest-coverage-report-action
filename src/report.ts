import core from '@actions/core';
import { ZodError } from 'zod';

import { getCurrentBranch } from './git';
import { getOptions } from './options';

export interface InvalidOptionsReport {
    type: 'invalid-options';
    error: ZodError;
}

export type CovbotReport = InvalidOptionsReport;

export const getReport = async (): Promise<CovbotReport> => {
    core.debug('[stage]: getting options');
    const options = getOptions();

    /**
     * Report error if failed to parse options.
     * There is nothing more we could do.
     */
    if (!options.success) {
        return {
            type: 'invalid-options',
            error: options.error,
        };
    }

    /**
     * Try to get current branch on current repository.
     * This is needed just for consistency - when action ends work, return to initial branch.
     * It will automatically report warning, if getting this branch failed.
     */
    core.debug('[stage]: getting current branch');
    const currentBranch = await getCurrentBranch();

    /**
     * If it is pull request, or pull request number is present, get PR info
     */
    const pullRequest = getPullRequestInfo();
};
