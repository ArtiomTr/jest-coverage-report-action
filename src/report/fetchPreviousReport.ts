import { getOctokit } from '@actions/github';

import { getReportTag } from '../constants/getReportTag';
import { Options } from '../typings/Options';

export async function fetchPreviousReport(
    octokit: ReturnType<typeof getOctokit>,
    repo: { owner: string; repo: string },
    pr: { number: number },
    options: Options
) {
    const commentList = await octokit.paginate(
        'GET /repos/{owner}/{repo}/issues/{issue_number}/comments',
        {
            ...repo,
            issue_number: pr.number,
        }
    );

    const previousReport = commentList.find((comment) =>
        comment.body?.startsWith(getReportTag(options))
    );

    return !previousReport ? null : previousReport;
}
