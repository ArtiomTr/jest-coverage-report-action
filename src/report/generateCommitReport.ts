import { context, getOctokit } from '@actions/github';

export const generateCommitReport = async (
    report: string,
    repo: { owner: string; repo: string },
    octokit: ReturnType<typeof getOctokit>
) => {
    await octokit.repos.createCommitComment({
        ...repo,
        commit_sha: context.sha,
        body: report,
    });
};
