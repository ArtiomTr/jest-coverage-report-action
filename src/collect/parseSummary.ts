import {
    coveredBranchesCounter,
    coveredLinesCounter,
    standardCoveredCounter,
    standardTotalCounter,
    totalBranchesCounter,
    totalLinesCounter,
} from './counters';
import { getSummary } from './getSummary';
import { JsonReport } from '../typings/JsonReport';

export const parseSummary = (jsonReport: JsonReport) => {
    return [
        getSummary(
            jsonReport.coverageMap,
            standardTotalCounter('s'),
            standardCoveredCounter('s'),
            'Statements'
        ),
        getSummary(
            jsonReport.coverageMap,
            totalBranchesCounter,
            coveredBranchesCounter,
            'Branches'
        ),
        getSummary(
            jsonReport.coverageMap,
            standardTotalCounter('f'),
            standardCoveredCounter('f'),
            'Functions'
        ),
        getSummary(
            jsonReport.coverageMap,
            totalLinesCounter,
            coveredLinesCounter,
            'Lines'
        ),
    ];
};
