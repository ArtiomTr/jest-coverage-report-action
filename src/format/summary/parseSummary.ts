import { getSummary } from './getSummary';
import { JsonReport } from '../../typings/JsonReport';
import { i18n } from '../../utils/i18n';
import {
    coveredBranchesCounter,
    coveredLinesCounter,
    standardCoveredCounter,
    standardTotalCounter,
    totalBranchesCounter,
    totalLinesCounter,
} from '../counters';

export const parseSummary = (jsonReport: JsonReport) => {
    return [
        getSummary(
            jsonReport.coverageMap,
            standardTotalCounter('s'),
            standardCoveredCounter('s'),
            i18n('statements')
        ),
        getSummary(
            jsonReport.coverageMap,
            totalBranchesCounter,
            coveredBranchesCounter,
            i18n('branches')
        ),
        getSummary(
            jsonReport.coverageMap,
            standardTotalCounter('f'),
            standardCoveredCounter('f'),
            i18n('functions')
        ),
        getSummary(
            jsonReport.coverageMap,
            totalLinesCounter,
            coveredLinesCounter,
            i18n('lines')
        ),
    ];
};
