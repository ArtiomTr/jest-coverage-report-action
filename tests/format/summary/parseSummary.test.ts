import { parseSummary } from '../../../src/format/summary/parseSummary';
import report from '../../mock-data/jsonReport.json';

describe('parseSummary', () => {
    it('should parse summary from mock data', () => {
        expect(parseSummary(report)).toMatchSnapshot();
    });
});
