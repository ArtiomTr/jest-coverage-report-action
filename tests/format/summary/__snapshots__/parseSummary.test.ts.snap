// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseSummary > should parse summary from mock data 1`] = `
[
  {
    "covered": 27,
    "percentage": 81.81818181818183,
    "title": "Statements",
    "total": 33,
  },
  {
    "covered": 8,
    "percentage": 100,
    "title": "Branches",
    "total": 8,
  },
  {
    "covered": 7,
    "percentage": 63.63636363636363,
    "title": "Functions",
    "total": 11,
  },
  {
    "covered": 25,
    "percentage": 80.64516129032258,
    "title": "Lines",
    "total": 31,
  },
]
`;
